### ISSUE TITLE
Problems with locale and external applications on the cmd.-line

### ISSUE THREAD

**Issue by WolfgangMehner** [neovim/neovim#9886]
- `nvim --version`: v0.4.0-575-gb358054
- Vim (version: ) behaves differently? yes, with 8.1 patches 1-1115
- Operating system/version: Ubuntu 16.04
- Terminal name/version: GNOME-terminal and Konsole (from KDE)
- `$TERM`: xterm-256color and xterm, respectively

### Steps to reproduce using `nvim -u NORC`

```
nvim -u NORC

:let $LANG='en'   " or whatever is an invalid locale

:! evince somedoc.pdf     " (1) blocking
:! evince somedoc.pdf &   " (2) or non-blocking

```

### Actual behaviour

Alternative (1) blocks the editor, `evince` starts, and a warning is displayed on the nvim cmd.-line: "Locale not supported by C library. Using the fallback 'C' locale."

Alternative (2) does not work, and does not display a warning. The external program does not start up.

_Edit_: When running inside a terminal in Neovim, started via `:terminal`, both shell commands work.

Some applications can be started using `:!<app>` and `:!<app> &` alike, some can not. Both alternatives work with:
- git gui
- gitk
- nvim-qt

The non-blocking alternative (2) does not work with:
- cmake-gui
- doxywizard
- evince
- okular
- geeqie

### Expected behaviour

Both alternatives should start the external application, even though the locale is wrong. They do both work in `vim` and `gvim`.

**Comment by justinmk** [comment]
> When running inside a terminal window, both shell commands work.

 `:terminal` starts an interactive shell, maybe your shell  inits various env vars (check `env` in the `:terminal` shell to confirm)?

What are you requesting exactly? Nvim cannot make the decision to fixup env vars that were explicitly set...

**Comment by WolfgangMehner** [comment]
To clarify, with "inside a terminal window" I mean a terminal inside Neovim started via `:terminal`. This was quite ambiguous, sorry.

What I'm requesting is that both version __(1)__ and __(2)__ work, as they do in Vim/gVim. The problem with the localization should not matter, since the application works when started by any other means. Additionally, the env vars don't really need fixing, the external application should and does fall back to some default.

Somehow the addition of the `&` to run the command in the background leads to a radically different behavior. I realize that this might be a nasty interaction between the way Neovim forks out the command and the external application, but Vim manages to deal with this. 

This is undesirable, since this has important usecases. When editing tex or markdown, one might what to map commands like these:
```
:! evince %:r.pdf &
:! firefox %:r.html &
```

**Comment by justinmk** [comment]
> What I'm requesting is that both version (1) and (2) work, as they do in Vim/gVim

That isn't concrete though. 

>  interaction between the way Neovim forks out the command and the external application, but Vim manages to deal with this.

Vim deals with it by accident. Nvim always invokes processes in a separate "session id" / process-group. Unless there is something concrete that Nvim is doing wrong, making things match some accidental effect is just not actionable.

**Comment by WolfgangMehner** [comment]
The problem is not really the interactive nature of `!` or lack thereof, that just coincidentally fixes a problem with the external application. `evince` can't cope if `stderr` is not connected. Running in a shell, outside Neovim. this also fails to start the application, or more specifically its GUI:

    LANG='en' evince bash-hotkeys.pdf 2> /notavalidfile

On the other hand, this works:
```
nvim -u NORC

:let $LANG='en'
:! evince somedoc.pdf 2> /dev/null &
```

So the problem seems to be this. Alt __(1)__ connects `stderr` correctly and display the error message on the Nvim cmd.-line:
```
:! evince somedoc.pdf

(process:7641): Gtk-WARNING **: Locale not supported by C library.
        Using the fallback 'C' locale.
```

I guess Alt __(2)__ returns to Nvim immediately, and the pipe Nvim opened for the shell command is closed. Then `evince` tries to write to `stderr` and fails, thus never starting its GUI. This would happen with any sort of warning message printed by the external application, the locale issue just happened to be the one I got.
```
:! evince somedoc.pdf &
```

I can see now how this derives from the new way Neovim handles the `:!` command with pipes. In the very least, the documentation should state more clearly that `:!` can't be used to start external applications in the background which may still depend on their `stdout` and `stderr` being writable. The documentation should clarify whether or not it is now the responsibility of the user to deal with this. Linking the documentation for `:!` to `jobstart` might also help.

Should I write a consolidated report for the thread #1496, so that all problems related to the new `:!` are listed there?

**Comment by WolfgangMehner** [comment]
TL;DR

Since this came up before at least once #6564, I'll just make my most important point in isolation:

In the very least, the documentation should state more clearly that `:!` can't be used to start external applications in the background which may still depend on their stdout and stderr being writable. The documentation should clarify whether or not it is now the responsibility of the user to deal with this. Linking the documentation for `:!` to `jobstart` might also help.

**Comment by justinmk** [comment]
Just to be clear, does `jobstart()` (perhaps with certain parameters) and/or `:terminal` work around the issue? This is something we can mention in the docs.

> the documentation should state more clearly that `:!` can't be used to start external applications in the background which may still depend on their `stdout` and `stderr` being writable.

We can add a note about that detail. But the story is the same as always since #1496: "to run an interactive program, use `jobstart()` or `:terminal`". evince is an interactive program.

**Comment by WolfgangMehner** [comment]
This does exactly what I want:
```
call jobstart ( '<cmd>', { 'detach' : 1 } )
```
The terminal is of course not a good option where since I just want to start the process, but it also works.

In this context, since people are mostly discussing terminals, I understood interactive as a text-only program running a REPL. Your definition is obviously wider. I have a feeling I might not be the only one getting confused here. Usecases like `git add -p` and starting an external program with a GUI probably seem quite different to many users.

**Comment by justinmk** [comment]
see https://github.com/neovim/neovim/pull/9946
